core/src/connection.rs

```diff
@@ -19,27 +19,37 @@
 // DEALINGS IN THE SOFTWARE.
 
 use crate::multiaddr::{Multiaddr, Protocol};
+use std::fmt;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+static NEXT_CONNECTION_ID: AtomicU64 = AtomicU64::new(0);
 
 /// Connection identifier.
-#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
-pub struct ConnectionId(usize);
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ConnectionId(u64);
 
 impl ConnectionId {
     /// Creates a `ConnectionId` from a non-negative integer.
     ///
     /// This is primarily useful for creating connection IDs
     /// in test environments. There is in general no guarantee
     /// that all connection IDs are based on non-negative integers.
+    #[deprecated(note = "IDs must be unique and should not be constructed directly.")]
     pub fn new(id: usize) -> Self {
-        Self(id)
+        Self(id as u64)
     }
 }
 
-impl std::ops::Add<usize> for ConnectionId {
-    type Output = Self;
+impl Default for ConnectionId {
+    fn default() -> Self {
+        Self(NEXT_CONNECTION_ID.fetch_add(1, Ordering::SeqCst))
+    }
+}
 
-    fn add(self, other: usize) -> Self {
-        Self(self.0 + other)
+impl fmt::Display for ConnectionId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
     }
 }
```

Comment:
TODO: Wait for https://github.com/libp2p/rust-libp2p/pull/3221 to be merged here.